src/Shared/runtime/Http2/Hpack/HPackDecoder.cs

```diff
@@ -492,23 +493,36 @@ private void CheckIncompleteHeaderBlock(bool endHeaders)
 
         private void ProcessHeaderValue(ReadOnlySpan<byte> data, IHttpHeadersHandler handler)
         {
-            ReadOnlySpan<byte> headerNameSpan = _headerNameRange == null
-                ? new Span<byte>(_headerName, 0, _headerNameLength)
-                : data.Slice(_headerNameRange.GetValueOrDefault().start, _headerNameRange.GetValueOrDefault().length);
-
             ReadOnlySpan<byte> headerValueSpan = _headerValueRange == null
                 ? new Span<byte>(_headerValueOctets, 0, _headerValueLength)
                 : data.Slice(_headerValueRange.GetValueOrDefault().start, _headerValueRange.GetValueOrDefault().length);
 
-            handler.OnHeader(headerNameSpan, headerValueSpan);
-
-            _headerNameRange = null;
-            _headerValueRange = null;
+            if (_headerStaticIndex > 0)
+            {
+                handler.OnStaticIndexedHeader(_headerStaticIndex, headerValueSpan);
 
-            if (_index)
+                if (_index)
+                {
+                    _dynamicTable.Insert(H2StaticTable.Get(_headerStaticIndex - 1).Name, headerValueSpan);
+                }
+            }
+            else
             {
-                _dynamicTable.Insert(headerNameSpan, headerValueSpan);
+                ReadOnlySpan<byte> headerNameSpan = _headerNameRange == null
+                    ? new Span<byte>(_headerName, 0, _headerNameLength)
+                    : data.Slice(_headerNameRange.GetValueOrDefault().start, _headerNameRange.GetValueOrDefault().length);
+
+                handler.OnHeader(headerNameSpan, headerValueSpan);
+
+                if (_index)
+                {
+                    _dynamicTable.Insert(headerNameSpan, headerValueSpan);
+                }
             }
+
+            _headerStaticIndex = 0;
+            _headerNameRange = null;
+            _headerValueRange = null;
```

Comment:
Just as idea:
Separate this out in a struct, so the clearing can be done just by `= default`, and the JIT will [emit more efficient code](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHuGAHKAJYA3bBhgMBAOwwMA+gAsY2ACYwoAZQyiBYAJKTVCANxce/YaPEAKKTNxaoGNBOkMANjEkBzDPICUAPxyiipqAHLY+DAAStjeMCbUPHyCImIMNq722I7Otu6ePv5BCkqqUABq2G4ArjFxXgmm3M1MjMQoDACyVn4crcmloRpaGDr6hgwAvAwADInJgyHlEVGx8dMMkjVubguL3EPlVbX1GzPbu/vcAL40d9Q09EykDAAi/UlmqZYM9lA1MAyN6ibCfA7cZ75AASZTUmm0egMMGMAx4z0ydgcThcMg83l8gQYsOGqzOjWuiwx+WyuVxBQJxWJcMq1Tq6wprQeyVa5jS4hBWjkylBlNazw63V64IhshFQpmqgAZtgdhhKQ8bkA==).
Codegen for individual field accesses should remain in the same machine code.

I have no feeling how hot this method is, so don't know if this it worth it.

Reply:
I want to leave low-level optimizations like that out of this PR, but you're welcome to take a look at HPackDecoder. There are probably a lot of places where small improvements (e.g. eliminating bounds checks, inlining methods)  can be made.

Reply:
Sounds good. Will have a look after this PR is in.